### PR TITLE
Support for custom conditions before aggregates are locked

### PR DESCRIPTION
**Pull request by hilkeheremans**
Hey @adrai, see #75  for the reasoning behind this PR. There is of course no obligation to accept, but I thought it would be more convenient to submit a PR so you can easily review and comment on the changes.

(Might I note it is also rather hard to run the full test suite due to the dependencies on the databases? Some instructions on what to install / pointers to docker containers with the proper dependencies would be great.)

### CODE REVIEW

File: lib/defaultCommandHandler.js [hunk 1]
```diff
@@ -775,6 +779,12 @@ _.extend(DefaultCommandHandler.prototype, {
       // lock aggregate
```
Comment (adrai) [inline lib/defaultCommandHandler.js]: wrong comment?

File: lib/defaultCommandHandler.js [hunk 2]
```diff
@@ -775,6 +779,12 @@ _.extend(DefaultCommandHandler.prototype, {
...
       function (clb) {
         debug('lock aggregate');
```
Comment (adrai) [inline lib/defaultCommandHandler.js]: wrong debug?

File: lib/definitions/preCondition.js [hunk 3]
```diff
@@ -30,7 +30,7 @@ function PreCondition (meta, preConditionFn) {
   this.payload = meta.payload || null;
   this.priority = meta.priority || Infinity;
 
-  this.preConditionFn = preConditionFn;
+  this.preLoadConditionFn = preConditionFn;
```
Comment (adrai) [inline lib/definitions/preCondition.js]: This is the preCondition file... why this change?

File: lib/definitions/preCondition.js [hunk 4]
```diff
@@ -89,16 +89,16 @@ _.extend(PreCondition.prototype, {
     }
 
     try {
-      if (this.preConditionFn.length === 3) {
-        this.preConditionFn(_.cloneDeep(payload), aggData, function (err) {
+      if (this.preLoadConditionFn.length === 3) {
```
Comment (adrai) [inline lib/definitions/preCondition.js]: This is the preCondition file... why this change?

File: lib/definitions/preCondition.js [hunk 5]
```diff
@@ -89,16 +89,16 @@ _.extend(PreCondition.prototype, {
...
+        this.preLoadConditionFn(_.cloneDeep(payload), aggData, function (err) {
           if (err) {
             return handleError(err);
           }
           callbacked = true;
           callback(null);
         });
       } else {
-        var err = this.preConditionFn(_.cloneDeep(payload), aggData);
+        var err = this.preLoadConditionFn(_.cloneDeep(payload), aggData);
```
Comment (adrai) [inline lib/definitions/preCondition.js]: This is the preCondition file... why this change?

**Comment by hilkeheremans** [comment]
Sorry. It was a late night change. I'll fix things up tonight and re-submit a PR. Keep 'm coming :-)

**Comment by adrai** [comment]
Every contribution to this cqrs modules is a benefit.
And this is really a great birthday present ;-)

I'm going to merge this as soon as possible...
Just a few questions:
It's completely backwards compatible, right? So minor release should be ok?
Just to know the reason, but why is the aggregate not locked before running the preLoadConditions? Why it's so important not to load the aggregate first? (Probably THE reason why not to use preConditions)
Some comments in the code...

File: npm-debug.log [hunk 6]
```diff
@@ -0,0 +1,48 @@
+0 info it worked if it ends with ok
```
Comment (adrai) [inline npm-debug.log]: npm-debug.log => remove and .gitignore?

File: releasenotes.md [hunk 7]
```diff
@@ -1,3 +1,109 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
```
Comment (adrai) [inline releasenotes.md]: I prefer not to have a TOC in releasenotes...

**Comment by hilkeheremans** [comment]
Yes, it's backwards compatible as far as I can test.

The reason for introducing this: assume you have a situation where you want to be able to process orders at a face pace (eg multiple per second). You want to check each order for some validity (are the articles in the basket valid, is the user allowed to order or otherwise not allowed to order for some reason, ...).

It is possible to check this by enlarging the aggregate accordingly and using preconditions. However, the aggregate then becomes so large that you will introduce a bottleneck, both data-wise as well as locking-wise, since on each order placement you'll need to lock the aggregate. An alternative would be to use saga's to handle these checks, but this introduces a lot of extra commands and events that are passed around on the bus. Using the custom command handler probably works too, but is cumbersome and prone to error.

I'll add some context to the docs and in the code in the next, proper PR.

**Comment by adrai** [comment]
super, thx ;-)

**Comment by hilkeheremans** [comment]
This should help. Let me know what other remarks you might have.

**Comment by adrai** [comment]
v2.3.0
